### DISCUSSION TITLE
User defined types in Risor

### DISCUSSION THREAD

**Discussion by myzie** [deepnoodle-ai/risor#138] [Ideas]
Risor needs the ability for users to define their own types, with associated methods and/or data. All the specifics for this are up for discussion, and we could draw inspiration from Go structs and interfaces, Typescript interfaces, or types in other languages including Rust, Zig, Python, Lua, etc.

Please share your ideas!

**Comment by applejag** [comment]
Just syntactically, having optional types on declarations should work just fine, with just reusing the Go syntax, such as:

```go
// Implicit declaration
var x = [1, 2, 3] // implicit literal []any

// Explicit declaration
var x []int = [1, 2, 3] // implicit literal turning into []int
```

But probably the bigger question is: does the Risor syntax allow optional Go-like type declarations on literals?

À la:

```go
x := [1, 2, 3] // implicit []any

x := []int[1, 2, 3] // explicit []int
```

This will create a conflict when parsing. Without looking ahead while parsing, it's ambiguous if the `[]` part is an empty slice or the start of a type. Having the parser depend on looking ahead is a bad idea, as it can overly complicate the parser logic. Especially when considering supporting TextMate and tree-sitter grammars.

Adding Go's type syntax as-is is probably not possible.

Maybe Risor would be OK without types on literals but instead only defining the type on the variable (`var x []int = ...`), and only relying on type coercion during compilation?

If type declaration on literals (`x := []int[1, 2, 3]`) is also wanted, then I strongly suggest a different syntax than Go's syntax

**Reply by myzie** [reply]
The Risor parser does "peek" at the next token already in some cases, when deciding how to parse the source. For example set and map literals are ambiguous at the opening `{`. I'm supportive of optional types on declarations, and still lean towards using Go's syntax for this if possible.

**Comment by applejag** [comment]
Handling the "method receiver"/"`this` variable"/"self" is an interesting design issue. In Go it looks like this:

```go
func (m MyType) MyFunc() {}
```

How should Risor handle this?

LUA handles this by dedicating the first parameter as the "self" variable, and giving you special syntax when calling. Meaning you as a Risor programmer need to keep track of this.

```lua
local function my_func() end
local my_type = {my_func = my_func}

-- These two are semantically identical
my_type.my_func()
my_func()

-- These two are semantically identical
my_type:my_func() -- notice the ":"
my_func(my_type)
```

It's a really simple approach, and is the simplest to implement. Quite tedious though, and can be a little hard to keep remembering when learning the language.

JavaScript tries to handle this all behind the scenes by having an internal variable called `this` that is referring to the caller. The only way to change this variable is when calling a function, or when using the `.bind()` function on a function.

```js
function myFunc() {}
class MyClass {}
MyClass.prototype.myFunc = myFunc;
const myObj = new MyClass();

// These two are semantically identical
myObj.myFunc();
myFunc.bind(myObj)();
```

Personally I like the JavaScript approach.

A caveat with JavaScript's approach is that it makes using anonymous functions trickier. They have an escape hatch: when using arrow function notation, it does not change the `this` variable:

```js
const myType = {
  x: 123,
  foo: function() {
    console.log('this.x in foo:', this.x);
    const anon = function() {
      console.log('this.x in anonymous function:', this.x);
    };
    const arrow = () => {
      console.log('this.x in arrow function:', this.x);
    };
    anon();
    arrow();
  },
};

myType.foo();
```

```console
$ node example.js
this.x in foo: 123
this.x in anonymous function: undefined
this.x in arrow function: 123
```

This is super annoying, and can lead to some annoying bugs when you're not careful. Especially when you don't have a strong language server (where TypeScript is saving JavaScript programmer's sanity).

However, we can make this super transparent to the programmer if we borrow the syntax from Go, as then you don't fight over the same `this` variable. Instead, you always explicitly declare when you're interested in the `this` variable.

What I propose is that Risor gets the same internal `this` variable that is tracked, but instead from a hidden variable. For the sake of argument, lets imagine this Risor code:

```go
func (m MyType) my_func() {
  print(m.x)
}
```

was just syntactic sugar for this:

```go
MyType.my_func = func() {
  m := this // "this" here is a hidden variable that only the compiler has access to
  print(m.x)
}
```

So that there would be an op-code to load in the current `this` variable on to the value stack and then load that into the variable `m` in the beginning of the function.

**Comment by SolarLune** [comment]
An idea that's similar to the JS method in applejag's approach would be to just use maps as objects. Then you could define functions as first-class objects that can refer to the map outside of its scope:

```go

func NewCar(topSpeed) {

    car := {
        wheels : 4,
        speed : 0,
        topSpeed : topSpeed,
    }

    // car.drive is just another way to access car["drive"].
    // This isn't necessary, of course, but it's nice to look at
    car.drive = func() {
        car.speed = min(car.speed + 1, car.topSpeed)
    }

    return car

}

stationWagon := NewCar(120)

ferrari := NewCar(270)

```

You could make "class-level static constants" by just referring to variables outside of the object / method definitions in the constructor.

But I think I like Applejag's Go-like approach more.

EDIT 2: To expand on this a bit with a little untested wonky quasi-inheritance:

```go

func NewSupercar() {

    car := NewCar(300)

    ogDrive := car.drive

    // Overload car.drive(), but refer to and call the original drive() function
    car.drive = func() {
        ogDrive()
        car.speed *= 1.315
    }

    return car

}


```

**Reply by rubiojr** [reply]
@SolarLune I've been using this pattern to wrap an HTTP API (in https://github.com/rubiojr/tlz) and I'm enjoying it very much. Quite clever!
